com.ibm.wala.core/src/main/java/com/ibm/wala/util/scope/JUnitEntryPoints.java

```diff
@@ -44,33 +67,85 @@
     }
+
     final HashSet<Entrypoint> result = HashSetFactory.make();
     for (IClass klass : cha) {
-      if (klass.getClassLoader().getReference().equals(ClassLoaderReference.Application)) {
+      IClassLoader classLoader = klass.getClassLoader();
+      ClassLoaderReference reference = classLoader.getReference();
+      if (reference.equals(ClassLoaderReference.Application)) {
         // if the class is a subclass of the Junit TestCase
         if (isJUnitTestCase(klass)) {
-
-          System.out.println("application class: " + klass);
+          logger.fine(() -> "application class: " + klass);
 
           // return all the tests methods
           Collection<? extends IMethod> methods = klass.getAllMethods();
 
           for (IMethod m : methods) {
             if (isJUnitMethod(m)) {
               result.add(new DefaultEntrypoint(m, cha));
-              System.out.println("- adding test method as entry point: " + m.getName().toString());
+              logger.fine(() -> "- adding test method as entry point: " + m.getName().toString());
             }
           }
+
+          // add entry points of setUp/tearDown methods
+          Set<IMethod> setUpTearDowns;
+          try {
+            setUpTearDowns = getSetUpTearDownMethods(klass);
+          } catch (Exception e) {
+            throw new IllegalArgumentException(
+                "Can't find test method entry points using class hierarchy: " + cha, e);
+          }
+          for (IMethod m : setUpTearDowns) {
+            result.add(new DefaultEntrypoint(m, cha));
+          }
+        } else { // JUnit4?
+          boolean isTestClass = false;
+
+          // Since JUnit4 test classes are POJOs, look through each method.
+          for (com.ibm.wala.classLoader.IMethod method : klass.getDeclaredMethods()) {
+            // if method has an annotation
+            if (!(method instanceof ShrikeCTMethod)) continue;
+            for (Annotation annotation : ((ShrikeCTMethod) method).getAnnotations())
+              if (isTestEntryPoint(annotation.getType().getName())) {
+                result.add(new DefaultEntrypoint(method, cha));
+                isTestClass = true;
+              }
+          }
+
+          // if the class has a test method, we'll also need to add it's ctor.
+          if (isTestClass) {
+            IMethod classInitializer = klass.getClassInitializer();
+
+            if (classInitializer != null) result.add(new DefaultEntrypoint(classInitializer, cha));
+
+            IMethod ctor = klass.getMethod(MethodReference.initSelector);
+
+            if (ctor != null) result.add(new DefaultEntrypoint(ctor, cha));
+          }
         }
       }
     }
     return result::iterator;
   }
 
+  private static boolean isTestEntryPoint(TypeName typeName) {
+    String javaName = walaTypeNameToJavaName(typeName);
+    return TEST_ENTRY_POINT_ANNOTATION_NAMES.contains(javaName);
+  }
+
+  private static String walaTypeNameToJavaName(TypeName typeName) {
```

Comment:
Can we replace this method with just using `com.ibm.wala.util.strings.StringStuff.jvmToReadableType()` instead?

Reply:
Sounds plausible to me.